### DISCUSSION TITLE
Compile errors following Brian’s tutorial on developer dao workshop

### DISCUSSION THREAD

**Discussion by Ekemini004** [aptos-labs/aptos-developer-discussions#335] [Dev Doc and Tutorial Feedback]
Hi everyone please ido if it’s possible to find the Brian that held the tut on developer dao 

been trying to run my code and it’s not getting past compile stage 

I’m getting sooo many errors - new to move tho 

watched the tutorial and made something similar, the exact same thing he wrote but it didn’t compile 

do I share the code or ?🥹I really need answers

**Comment by 0xmaayan** [comment]
Hi @Ekemini004 , could you provide the steps to reproduce the error? as well as the actual errors you are getting

**Reply by Ekemini004** [reply]
This is my code :
// An application where every user has a name and an XP counter
// Security consideration: Only the user should be able to change their name or XP
module ekems_aptos1::Xp_counter {
    //Importing neccesary dependencies -- idrk why
   use aptos_framework::event;
   use std::signer::{address_Of};
   use std::string::{String, utf8}; 

  // Define a struct called Account to hold user name and xp
    struct Account has key {
       name: String,
       xp: Xp
    }

  // Define a struct called Xp to hold Count
    struct Xp has store {
      count: u64,
        
    }
       // Emit an event
       #[event] 
       struct XpChanged has drop,store {
        user_addr: address,
        new_xp: u64,
        }

         #[event]
       struct NameChanged has drop,store {
        user_addr: address,
        new_name: String,
        }

        //Account has not been initialized yet
        //This is the function that initializes the account upon signup
        const ENOT_INITIALIZED: u64 = 0;
//function visibility by default is private, so using fun function name makes the function private
//adding the public keyword makes the function explicitly public, in that other smart contracts can call the function, but when a user signs a transcation he/she cant call a function that has the public keyword infront of it
//so public keyword just makes functions public to other smart contracts but not public enough for a user or transcation signer to call 
//adding the entry keyword makes the function visible enough to be called by a transaction signer and other smart contracts
        public entry fun init_account(signer: &signer, name: String) {
          move_to(signer, Account { name: name, xp: Xp { count: 0 } });
        }

      //function to increase  xps
      public entry fun increase_Xp(user: &signer, amount: u64) acquires Account {
        let user_addr = address_Of(user); 
        assert!(account_exists(user_addr), ENOT_INITIALIZED);
        let account = borrow_global_mut<Account>(user_addr);
        let new_xp = account.xp.count + amount;
        account.book.count = new_xp;
        event::emit(XpChanged{user_addr, new_xp});


      }

      //function to reduce xps
       public entry fun decrease_Xp(user: &signer, amount: u64) acquires Account {
        let user_addr = address_Of(user); 
        assert!(account_exists(user_addr), ENOT_INITIALIZED);
        let account = borrow_global_mut<Account>(user_addr);
        let new_xp = account.xp.count - amount;
        account.book.count = new_xp;
        event::emit(XpChanged{user_addr, new_xp});


      }


        //function to set name
        public entry fun set_name(user: &signer, new_name: String) acquires Account
        {
          let user_addr = address_Of(user);
          assert!(account_exists(user_addr), ENOT_INITIALIZED);
          let account = borrow_global_mut<Account>(user_addr);
          account.name = new_name;
          event::emit(NameChanged{user_addr, new_name});
          }

          //view functions enable us check on the explorer to see details of our state

      #[view]
      public fun account_exists(addr: address): bool{
        exists<Account>(addr)
      }

      #[view]
      public fun get_xp(addr: address):u64 acquires Account{
        if (account_exists(addr)){
          borrow_global<Account>(addr).xp_count
        }else{
          0
        }
      }
     
     #[view]
     public fun get_name(addr: address): String acquires Account{
      if (account_exists(addr)){
        borrow_global<Account>(addr).name
        }else{
          utf8(b"")
          }}

          //writting a unit test code by mocking an admin with a fake address to test run the code
          #[test(admin = @0x12345)]
          
        public entry fun test_it_works(admin: &signer): acquires Account {
          assert!(!account_exists(address_Of(admin)), 0);

          init_account(admin, utf8(b"Ekemini"));
          assert!(get_xp(address_Of(admin)) == 0,0);
          assert!(get_name(address_Of(admin)) == utf8(b"Ekemini"), 0);
          

          increase_xp(admin, 10);
          assert!(get_xp(address_Of(admin)) == 10, 0);

          decrease_xp(admin, 5);
          assert!(get_xp(address_Of(admin)) == 5,0);

          set_name(admin, utf8(b"Sifon"));
          assert!(get_name(address_Of(admin)) == utf8(b"Sifon"), 0);

         }


       

 




 }

this is the error that comes up when i try to compile : BUILDING boilerplate_template      
error[E01002]: unexpected token
    ┌─ move\sources\XPcounter.move:106:57
    │
106 │         public entry fun test_it_works(admin: &signer): acquires Account {
    │                                                         ^^^^^^^^
    │                                                         │
    │                                                         Unexpected 'acquires'
    │                                                         Expected a type name

{
  "Error": "Move compilation failed: Compilation error"
}
PS C:\Users\LDO SYSTEMS\Desktop\aptosbuild\my-aptos-dapp>  also, its the exact same thing Brian wrote in the tutorial i wonder why its not working for me
![Screenshot 2024-07-23 002616](https://github.com/user-attachments/assets/e978a71b-922d-4099-a40d-e8a7c61c8bb8)
in the screenshot above, the move analyzer server thing doesn't work for me idkif that affects anything

**Reply by 0xmaayan** [reply]
could you try remove the `:` from `public entry fun test_it_works(admin: &signer): acquires Account`

**Reply by Ekemini004** [reply]
also tried @0xmaayan , i added "signer" before the "acquires" keyword and got these errors: error[E03003]: unbound module member
  ┌─ move\sources\XPcounter.move:6:22
  │
6 │    use std::signer::{address_Of};
  │                      ^^^^^^^^^^ Invalid 'use'. Unbound member 'address_Of' in module '(std=0x1)::signer'
  │
  ┌─ C:\Users\LDO SYSTEMS/.move\https___github_com_aptos-labs_aptos-core_git_mainnet\aptos-move/framework/aptos-framework\../aptos-stdlib\../move-stdlib\sources\signer.move:1:13
  │
1 │ module std::signer {
  │             ------ Module '(std=0x1)::signer' declared here

error[E03009]: unbound variable
   ┌─ move\sources\XPcounter.move:46:25
   │
46 │         let user_addr = address_Of(user);
   │                         ^^^^^^^^^^ Invalid function usage. Unbound variable 'address_Of'

error[E03010]: unbound field
   ┌─ move\sources\XPcounter.move:50:9
   │
50 │         account.book.count = new_xp;
   │         ^^^^^^^^^^^^ Unbound field 'book' in '(ekems_aptos1=0x123)::Xp_counter::Account'

error[E03009]: unbound variable
   ┌─ move\sources\XPcounter.move:58:25
   │
58 │         let user_addr = address_Of(user);
   │                         ^^^^^^^^^^ Invalid function usage. Unbound variable 'address_Of'

error[E03010]: unbound field
   ┌─ move\sources\XPcounter.move:62:9
   │
62 │         account.book.count = new_xp;
   │         ^^^^^^^^^^^^ Unbound field 'book' in '(ekems_aptos1=0x123)::Xp_counter::Account'

error[E03009]: unbound variable
   ┌─ move\sources\XPcounter.move:72:27
   │
72 │           let user_addr = address_Of(user);
   │                           ^^^^^^^^^^ Invalid function usage. Unbound variable 'address_Of'

   ┌─ move\sources\XPcounter.move:89:11
   │
89 │           borrow_global<Account>(addr).xp_count
   │           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Unbound field 'xp_count' in '(ekems_aptos1=0x123)::Xp_counter::Account'  

{
  "Error": "Move compilation failed: Compilation error"

**Reply by Ekemini004** [reply]
> could you try remove the `:` from `public entry fun test_it_works(admin: &signer): acquires Account`

did and got this - error[E03003]: unbound module member
  ┌─ move\sources\XPcounter.move:6:22
  │
6 │    use std::signer::{address_Of};
  │                      ^^^^^^^^^^ Invalid 'use'. Unbound member 'address_Of' in module '(std=0x1)::signer'
  │
  ┌─ C:\Users\LDO SYSTEMS/.move\https___github_com_aptos-labs_aptos-core_git_mainnet\aptos-move/framework/aptos-framework\../aptos-stdlib\../move-stdlib\sources\signer.move:1:13
  │
1 │ module std::signer {
  │             ------ Module '(std=0x1)::signer' declared here

error[E03009]: unbound variable
   ┌─ move\sources\XPcounter.move:46:25
   │
46 │         let user_addr = address_Of(user);
   │                         ^^^^^^^^^^ Invalid function usage. Unbound variable 'address_Of'

error[E03010]: unbound field
   ┌─ move\sources\XPcounter.move:50:9
   │
50 │         account.book.count = new_xp;
   │         ^^^^^^^^^^^^ Unbound field 'book' in '(ekems_aptos1=0x123)::Xp_counter::Account'

error[E03009]: unbound variable
   ┌─ move\sources\XPcounter.move:58:25
   │
58 │         let user_addr = address_Of(user);
   │                         ^^^^^^^^^^ Invalid function usage. Unbound variable 'address_Of'

error[E03010]: unbound field
   ┌─ move\sources\XPcounter.move:62:9
   │
62 │         account.book.count = new_xp;
   │         ^^^^^^^^^^^^ Unbound field 'book' in '(ekems_aptos1=0x123)::Xp_counter::Account'

error[E03009]: unbound variable
   ┌─ move\sources\XPcounter.move:72:27
   │
72 │           let user_addr = address_Of(user);
   │                           ^^^^^^^^^^ Invalid function usage. Unbound variable 'address_Of'

error[E03010]: unbound field
   ┌─ move\sources\XPcounter.move:89:11
   │
89 │           borrow_global<Account>(addr).xp_count
   │           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Unbound field 'xp_count' in '(ekems_aptos1=0x123)::Xp_counter::Account'  

{
  "Error": "Move compilation failed: Compilation error"
}
PS C:\Users\LDO SYSTEMS\Desktop\aptosbuild\my-aptos-dapp> also, does not hvain the move server for the move analyzer tension mean anything or could it contribute to the problem ?

**Reply by 0xmaayan** [reply]
I think `use std::signer::{address_Of};` should be `use std::signer::{address_of};`
https://github.com/aptos-labs/aptos-core/blob/main/aptos-move/framework/move-stdlib/doc/signer.md

**Reply by Ekemini004** [reply]
@0xmaayan corrected error on line 62->error[E03010]: unbound field
┌─ move\sources\XPcounter.move:62:9
│
62 │ account.book.count = new_xp;
│ ^^^^^^^^^^^^ Unbound field 'book' in '(ekems_aptos1=0x123)::Xp_counter::Account' but the rest , from my interaction with AI are issues with the address_Of and std::signer dependencies or modules. this code was written the exact same way brian did in the workshop, his worked but mine isnt , maybe there's something i don't have that he does ? could it be the move analyzer server?

**Reply by Ekemini004** [reply]
@Ekemini004 
![image](https://github.com/user-attachments/assets/3b96f514-b3c1-4c26-a9a1-b033b56185da)
i think this means it worked , thanks so much!!!

**Comment by Ekemini004** [comment]
hi @0xmaayan , been trying to create a petra wallet to test my functions on the explorer but it keeps hanging at confirm password and i don't even see the button to move to the next step
